Sources/ApolloCodegenLib/Glob.swift

```diff
@@ -0,0 +1,196 @@
+import Foundation
+
+private extension String {
+  static let Globstar: String = "**"
+
+  var includesGlobstar: Bool {
+    return self.contains(Self.Globstar)
+  }
+
+  var isExclude: Bool {
+    return self.first == "!"
+  }
+
+  var containsExclude: Bool {
+    return self.contains("!")
+  }
+}
+
+/// A path pattern matcher.
+public struct Glob {
+  let patterns: [String]
+
+  // GLOB_ERR - Return on error
+  // GLOB_MARK - Append / to matching directories
+  // GLOB_NOSORT - Don't sort
+  // GLOB_TILDE - Expand tilde names from the passwd file
+  private let flags = GLOB_ERR | GLOB_MARK | GLOB_NOSORT | GLOB_TILDE
+
+  /// An error object that indicates why pattern matching failed.
+  public enum MatchError: Error, LocalizedError, Equatable {
+    case noSpace // GLOB_NOSPACE
+    case aborted // GLOB_ABORTED
+    case cannotEnumerate(path: String)
+    case invalidExclude(path: String)
+    case unknown(code: Int)
+
+    public var errorDescription: String? {
+      switch self {
+      case .noSpace: return "Malloc call failed" // From Darwin.POSIX.glob
+      case .aborted: return "Unignored error" // From Darwin.POSIX.glob
+      case .cannotEnumerate(let path): return "Cannot enumerate \(path)"
+      case .invalidExclude(let path): return "Exclude paths must start with '!' - \(path)"
+      case .unknown(let code): return "Unknown error: \(code)"
+      }
+    }
+  }
+
+  /// The designated initializer
+  ///
+  /// - Parameters:
+  ///  - pattern: An array of path matching pattern strings.
+  ///
+  /// Each path matching pattern can include the following characters:
+  /// - `*` matches everything but the directory separator (shallow), eg: `*.graphql`
+  /// - `?` matches any single character, eg: `file-?.graphql`
+  /// - `**` matches all subdirectories (deep), eg: `**/*.graphql`
+  /// - `!` excludes any match only if the pattern starts with a `!` character, eg: `!file.graphql`
+  ///
+  /// - Discussion:
+  ///
+  init(_ patterns: [String]) {
+    self.patterns = patterns
+  }
+
+  /// Executes the pattern match on the underlying file system.
+  ///
+  /// - Returns: A set of matched file paths.
+  func match() throws -> Set<String> {
+    let expandedPatterns = try expand(self.patterns)
+
+    var includeMatches: [String] = []
+    var excludeMatches: [String] = []
+
+    for pattern in expandedPatterns {
+      if pattern.isExclude {
+        let patternMatches = try matches(for: String(pattern.dropFirst()))
+        excludeMatches.append(contentsOf: patternMatches)
+
+      } else {
+        let patternMatches = try matches(for: pattern)
+        includeMatches.append(contentsOf: patternMatches)
+      }
+    }
+
+    return Set<String>(includeMatches).subtracting(excludeMatches)
```

Comment:
Excludes are processed at the end and not in order - I think this is the better logic.